### ISSUE TITLE
LiveEffect issue on Android P

### ISSUE THREAD

**Issue by miqouser** [google/oboe#616]
I have an issue on latest Huawei P30 Pro (Android P). It hasn't 3.5mm Jack, but has USB Type-C headset, of course with microphone on it. The problem is I can't choose none of 2 "built in microphones" on device, it seems that library ignores that call, and headset's microphone continue it's work. The question is how to force my app to use "built in" microphone instead of headset's microphone. I even removed "AudioDeviceCallback" after all devices added, so headset isn't even listed, but it's working somehow.

Here is the list of devices (including USB headset too)

2019-08-27 17:51:09.591 9763-9763/? D/____all: built-in earphone speaker
2019-08-27 17:51:09.591 9763-9763/? D/____all: built-in speaker
2019-08-27 17:51:09.591 9763-9763/? D/____all: DOCK
2019-08-27 17:51:09.591 9763-9763/? D/____all: telephony
2019-08-27 17:51:09.591 9763-9763/? D/____all: built-in microphone
2019-08-27 17:51:09.591 9763-9763/? D/____what_I_need: built-in microphone
2019-08-27 17:51:09.591 9763-9763/? D/____all: USB device
2019-08-27 17:51:09.591 9763-9763/? D/____all: USB device

**Comment by philburk** [comment]
I tried to reproduce this on a Pixel 3a and it worked fine. The built-in mic was used for both listed mics.

It may be specific to the Huawei device.

@dturner - do you have a P30 Pro?

**Comment by miqouser** [comment]
Yes, on Pixel 3 it's fine, I've checked too. But here is a fact, with Google's Sound Amplifier I am able to choose microphone on mentioned device. Does it use AAudio engine too ? Maybe there is something with Huawei's headset ? Because with USB Type-C to 3.5mm Jack adapter I am able to choose microphone.

**Comment by dturner** [comment]
> @dturner - do you have a P30 Pro?

I'm afraid I don't. 

> But here is a fact, with Google's Sound Amplifier I am able to choose microphone on mentioned device.

Sound Amplifier actually uses Oboe I wonder how there is any difference in behaviour with LiveEffect. 

@miqouser Please could you submit a more detailed bug report with the following fields: 

Phone make and model and any other hardware used in the test
Android build
Steps to reproduce
Expected result
Actual result

It sounds like this might be a headset compatibility issue.

**Comment by miqouser** [comment]
Good afternoon, do you have any updates on this ? Thanks in advance.

**Comment by miqouser** [comment]
@dturner yes of course.

I have 3 Huawei devices: 

Mate 20 Pro (don't have 3.5mm Jack)
P30 Pro (don't have 3.5mm Jack)
P30 Lite (has 3.5mm Jack)

Also have multiple headsets and headphones, including original Huawei headsets.

Build is 9.1

Reproducing is simple, connect headphones with microphone (headset) and just try to change recording device.

Expected result is that recording device must be changed to AudioDeviceInfo.TYPE_BUILTIN_MIC for example (so works fine on any Pixel, Samsung, One Plus, Xiaomi etc).

Actual result is that recording device stays the same, i.e. headphone's microphone, seems app ignores setDeviceId function.

I have found 2 ways how to use Huawei built in microphones. One is to connect headphones without onboard microphone, then built in will be chosen, and the second way is to connect 3.5mm Jack headset (with microphone) via USB-TypeC adapter. 

Very very stranger things.

**Comment by philburk** [comment]
> app ignores setDeviceId function.

Are you selecting the OpenSL ES API? That is a known issue with OpenSL ES.
AAudio should honor the deviceID.

**Comment by miqouser** [comment]
No, I have completely removed OpenSL ES from my project. I am using only AAudio and minSdk version is above 27. So I must be able to choose recording device. On other devices that works fine.

**Comment by philburk** [comment]
@miqouser - OK, thanks for the detailed report.

Looks like a Huawei specific issue. I can get access to a Mate 20. Don and I will investigate and file a bug with Huawei.

**Comment by miqouser** [comment]
@philburk thank you too. Will wait for news. But as Don said:

**"Sound Amplifier actually uses Oboe I wonder how there is any difference in behaviour with LiveEffect."**

 Sound Amplifier works fine with Huawei and it's headset. I am sure the problem will be solved.

**Comment by miqouser** [comment]
@philburk sorry, have you any updates on this ?
